### ISSUE TITLE
Create gateway to Vertex

### ISSUE THREAD

**Issue by gmachtel** [nbh-digital/goldchain#31]
Gateway to NBHs Vertex, to enable to access Vertex back-office securely. 

Vertex is only secured by username & password. These credentials need to be kept and encrypted so that a 3bot user does not need to remember them. 

To do: create actors and connect to Vertex APIs. 

Vertex APIs 
[Login](https://support.hybridsolutions.com//index.php?/Knowledgebase/Article/View/1949
) (needed to authenticate user)
[NewOrder](https://hybrid-solutions.kayako.com//index.php?/Knowledgebase/Article/View/1952) (conversion GFT back to physical gold + (later phase) convert fiat <--> physical gold)
[GetAccounts](https://support.hybridsolutions.com//index.php?/Knowledgebase/Article/View/1934) (give overview of accounts with NBH IVS)
[GetAccountSummary](https://support.hybridsolutions.com//index.php?/Knowledgebase/Article/View/1953) (give content on a NBH IVS account)
[GetNewTick](http://www.hybrid-solutions.com/support/index.php?/Knowledgebase/Article/View/1939) (get a price)
[DeliverSymbol](url) (convert physical gold to GFTs)

**Comment by GlenDC** [comment]
@xmonader did some initial work purely in python related to their vertex API. It can be found at <https://gitlab.com/xmonader/vertex>. There are two more endpoints to finish off. It should give a headstart on this task.

The credentials required for their test environment as well as more information can be found in the Integration Telegram Group. @rkhamis is already part of it. Ask @gmachtel for more information on that group in if you need an invite.

Notes related to JSX integration:

- It will be implemented as a separate client (perhaps even two in case you are required to split their front- and backend clients in two JSX clients, otherwise if possible just one);
- It will be separate in the sense that it is not part of any blockchain client, but a standalone client;
- It will simply allow you to easily communicate with their servers using the provided and integrated vertex API endpoints, secured using the config-encrypted credentials (username, password for fontend, and there might be something else required for the backend endpoints, @gmachtel knows the details about that);
- The credentials can be part of the client JSX config, and marked as required to be encrypted, such that only the client's user can access it;
- The client won't be really that special and should be quite a routine task for any JSX developer, given that the vertex knowledge can be provided in the documentation and that most of it is already prepared in the repo provided by @xmonader;

Security is a big concern though:
- JSX is not the best currently in terms of security, so there is already that;
- the vertex API in itself is not that secure either:
  - username/password are passed as query parameters, while that might be "okay" given a secure connection, it might still give a leak in case they log their urls sever-side in plain text;
  - username/password combo with humans frequently result in security issues;
  - I do not know the full details, but it seems that according to @xmonader, certain code as part of a query or response is just evaluated as in the example, without any validation or security measurements;

Nothing is blocked here, and any experienced JSX developer should be able to resolve this issue within a couple of days, including testing and docs.

**Comment by robvanmieghem** [comment]
Can we add a spec of where which Vertex calls fit in to our flows?

**Comment by serboctor** [comment]
The specs have been changed. Only backend api calls will be needed. The `Gateway to Vertex` section here contains the relevant api calls needed https://docs.google.com/document/d/1pgBMoOjGRnNYIj4j7Xtq8XKqk1KrDCvTshODedSXjgc/edit#heading=h.a0fgutkov2u5

**Comment by serboctor** [comment]
vertex client is done and merged into `development` branch of jumpscaleX. I tested it and it works fine. https://github.com/threefoldtech/jumpscaleX/issues/786
The initial version of the gateway actors are pushed [here](https://github.com/nbh-digital/nbh-gateway). It's not complete. I am currently blocked on the implementation of the goldchain [client](https://github.com/nbh-digital/goldchain/issues/30) to add the rest of the business logic.

**Comment by serboctor** [comment]
I migrated the current actor to the new jumpscale and tested and fixed all the bugs. However, the implementation is still not complete as stated above.

**Comment by serboctor** [comment]
I added this README to explain what has been implemented so far.
https://github.com/nbh-digital/nbh-gateway/blob/master/gateway/actors/README.md

**Comment by serboctor** [comment]
Made some refactoring to the actors. We now filter on accounts that are not Locked before attempting to close positions on those accounts.

**Comment by serboctor** [comment]
Based on this message from @gmachtel, I'll pause until everything is settled.
```
Hi Sarah, I was busy with updating the flows. As we need to align again with both NBH Denmark and NBH Cyprus (Denmark would like to have a more simple system and not use Vertex for gold, but on the other side we need for eurotoken anyway a system that does positions), please leave it as it is. We'll come back to it after the discussions between Denmark and Cyprus are cleared out. Great job !```
